File: jpress-engine/src/main/java/by/ibragimov/jpress/generator/template/FreemarkerEngine.java
package by.ibragimov.jpress.generator.template;

import freemarker.cache.ClassTemplateLoader;
import freemarker.template.Configuration;
import freemarker.template.Template;
import freemarker.template.TemplateException;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.io.StringWriter;
import java.util.Map;
import java.util.function.Function;

/**
 * Used to convert "*.ftl" templates.
 *
 * @author Ibragimov Ruslan
 */
@Service
public class FreemarkerEngine implements TemplateEngine {

    private final Configuration configuration;

    public FreemarkerEngine() {
        configuration = new Configuration(Configuration.getVersion());
        configuration.setTemplateLoader(new ClassTemplateLoader(getClass(), "/"));
    }

    @Override
    public String render(final String template, final Map<String, Object> model) {
        try {
            Template freemarkerTemplate = configuration.getTemplate(template);
            StringWriter writer = new StringWriter();

            Function<String, String> function = (string) -> string;

            model.put("helper", function);

            freemarkerTemplate.process(model, writer);
            return writer.toString();
        } catch (IOException | TemplateException e) {
            throw new RuntimeException(e);
        }
    }
}

File: kpress-engine/src/main/kotlin/by/heap/kpress/generator/template/FreemarkerEngine.kt
package by.heap.kpress.generator.template

import freemarker.cache.ClassTemplateLoader
import freemarker.template.Configuration
import freemarker.template.TemplateException
import java.io.IOException
import java.io.StringWriter

/**
 * Used to convert "*.ftl" templates.
 *
 * @author Ruslan Ibragimov
 */
class FreemarkerEngine : TemplateEngine {
    private val configuration: Configuration = Configuration(Configuration.getVersion())

    init {
        configuration.templateLoader = ClassTemplateLoader(javaClass, "/")
    }

    override fun render(template: String, model: MutableMap<String, Any>): String {
        try {
            val freemarkerTemplate = configuration.getTemplate(template)
            val writer = StringWriter()

            val function = { string: String -> string }

            model["helper"] = function

            freemarkerTemplate.process(model, writer)
            return writer.toString()
        } catch (e: IOException) {
            throw RuntimeException(e)
        } catch (e: TemplateException) {
            throw RuntimeException(e)
        }
    }
}
